### ISSUE TITLE
Some Extra Parameters are not passing through to VRA

### ISSUE THREAD

**Issue by RoinGrindwald** [chef-partners/vmware-vra-gem#75]
### Versions:

* Version of vmware-vra-gem: **2.7.0**
* Version of chef: **14.10.9**
* Version of ruby: **2.5.0**

### Platform Details

* Version of vRA: **7.6**
* Version of ESXi: **6.5**

### Scenario:

Unable to pass all extra parameters through to VRA or once they are being validated during kitchen create, they are showing blank in VRA, even for required fields.

#### Kitchen.yaml
```
---
driver:
  name: vra
  username: 
  tenant: vsphere.local
  base_url: 
  verify_ssl: false

platforms:
  - name: Application Server Test
    driver:
      catalog_name: Application Basic
      subtenant_name: Windows Services
      cpus: 2
      memory: 4096
      notes: Testing
      extra_parameters:
        ppm_project_code:
          type: string
          value: 54321
        _deploymentName:
          type: string
          value: VRA Testing Builds - 7.6
        Hostname:
          type: string
          value: My-Test-Server01
        MaintenanceWindow:
          type: string
          value: SAT 12:00-13:00
        ServerDescription:
          type: string
          value: Testing Test Kitchen
        SRM:
          type: boolean
          value: false
        Vrm.DataCenter.Location:
          type: string
          values: Lab
        VirtualMachine.Network0.Name:
          type: string
          value: 212_VLAN

suites:
  - name: default
    run_list:
      - recipe[vra_kitchen_test::default]
    verifier:
      inspec_tests:
        - test/integration/default
    attributes:
```

### Expected Result:

Deployment would be sent to VRA to be built with all supplied parameters.

##### Error

```
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Failed to complete #create action: [Unable to submit request: The data specified within the request is invalid., Field with id ppm_project_code is required., Field with id SRM is required.] on default-Application Server Test
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```
It seems some parameters are not being sent through correctly or not at all.

#####  Parameter unable to send at all

```
      extra_parameters:
        ppm_project_code:
          type: string
          value: 54321
```

#####  Parameters that are validating but showing as blank in VRA on the request
```
      extra_parameters:
        _deploymentName:
          type: string
          value: VRA Testing Builds - 7.6
        Vrm.DataCenter.Location:
          type: string
          values: Lab
```

##### Parameter of boolean type not sending through

```
        SRM:
          type: boolean
          value: false
```

If changed to `string` it is able to pass-through

```
        SRM:
          type: string
          value: "false"
```
##### VRA Result

> Service Name = _deploymentName
![screenshot](https://user-images.githubusercontent.com/39578465/58436041-8a895e00-8117-11e9-9966-67a2fae100cd.png)

**Comment by stuartpreston** [comment]
TLDR: Please update vmware-vra to 2.7.1 to try out fixes for the issues seen.

I was able to reproduce the issues which seem to arise from a problem with the merging of parameters in the blueprint template with those supplied via kitchen.yml. The deployment properties at the top level should come through now:

My kitchen.yml:

```yaml
---
driver:
  name: vra
  username: stuart@vsphere.local
  tenant: vsphere.local
  base_url: https://vra.success.chef.co
  verify_ssl: false
  private_key_path: ~/.ssh/vagrant.pem

transport:
  username: vagrant

provisioner:
  name: chef_zero

verifier:
  name: inspec

platforms:
  - name: APP_Server
    driver:
      catalog_name: Application Basic
      subtenant_name: Windows Services
      cpus: 1
      memory: 2048
      notes: Notes go here
      extra_parameters:
        Deployment.PPMProjectCode:
          type: string
          value: A PPM Project Code goes here
        Deployment.ServiceName:
          type: string
          value: A Deployment Service Name goes here
        APP_Server~description:
          type: string
          value: A Description for the App Server goes here
        APP_Server~My.Domain:
          type: string
          value: A Domain here
        APP_Server~My.Network:
          type: string
          value: A Network Name here
        APP_Server~My.SRM:
          type: string
          value: A SRM here
        Deployment.AdditionalStringParameter:
          type: string
          value: A string value
        Deployment.AdditionalIntegerParameter:
          type: integer
          value: 999
        Deployment.AdditionalBooleanParameter:
          type: boolean
          value: true

suites:
  - name: default
    run_list:
```

Deployment:
![image](https://user-images.githubusercontent.com/6384654/58499802-ca346200-8178-11e9-9519-3348e319ba8f.png)

Component:
![image](https://user-images.githubusercontent.com/6384654/58500658-804c7b80-817a-11e9-83c4-ca4109b1d0d8.png)

I have added support (previously there was none) for passing through a parameter with a type of `boolean`.  All three types (String, Integer, Boolean) were accepted as additional parameters by vRA:

![image](https://user-images.githubusercontent.com/6384654/58499880-e506d680-8178-11e9-9ded-784164edb7c0.png)

The Deployment description field _should_ be set by the `notes` field rather than a parameter but doesn't appear to get set. This doesn't seem to be a problem at the component level, we will investigate this. I am not sure we pass through the Reason field either. 

I have also added support for tracing in the driver so that you can see the blueprint being requested as JSON and then the merged/submitted template being POSTed back. To enable this, set **VRA_HTTP_TRACE=1** in your environment prior to running Test Kitchen.

After this you will see every request and response hitting the vRA API. This can be useful to see the specific parameters being requested by the blueprint template.

![image](https://user-images.githubusercontent.com/6384654/58500540-51360a00-817a-11e9-8718-2f4f87c71ffb.png)

**Comment by RoinGrindwald** [comment]
Thanks @stuartpreston 

I have tested the new version and I have been able to get all the fields (Minus Description and Reason) to come through. Also, the addition of VRA_HTTP_TRACE is very much welcoming. I was able to see what was actually happening which helps debugging.

The only thing I notice which might still be a bug, is that if I add a mandatory property in the kitchen.yaml but fail to add a value to it, it will still send this request to VRA as NULL. This is invalid and should fail to send the request as a value is required.

```yaml
APP_Server~Vrm.DataCenter.Location:
  type: string
  value: 
```
![image](https://user-images.githubusercontent.com/39578465/58509341-6f3b5300-81ea-11e9-8854-4e69d7588937.png)

**Comment by stuartpreston** [comment]
@RoinGrindwald Thanks for the quick testing and glad to hear things are working. On your last point about mandatory fields I’m not sure there’s a lot we can do as the API doesn’t appear to provide a list of which fields are mandatory. 

I can’t think of a way around this without work to string parse the error message seen when you submit a null value for a mandatory field to get a list of these fields, do you have other ideas on how you’d see this working without more server round trips?

**Comment by RoinGrindwald** [comment]
@stuartpreston Hmm, I see what you mean. What you might need to do is one call to the Properties Service API of each Property Group found in the blueprint. Within the property group you will see mandatory value.

```json
"properties": {
    "_deploymentName": {
      "facets": {
        "visibility": {
          "type": "constant",
          "value": {
            "type": "boolean",
            "value": true
          }
        },
        "encrypted": {
          "type": "constant",
          "value": {
            "type": "boolean",
            "value": false
          }
        },
        "mandatory": {
          "type": "constant",
          "value": {
            "type": "boolean",
            "value": true
          }
        },
        "defaultValue": {
          "type": "constant",
          "value": {
            "type": "string",
            "value": ""
          }
        }
      }
    },
```

Also, is there a way to increase the kitchen create timeout value for deployments? If it gets to 600, if fails

Thanks
